csm_web/frontend/src/components/MentorSection.js

```diff
@@ -0,0 +1,328 @@
+                    onClick={() =>
+                      this.setState({ selectedWeek: weekStart, stagedAttendances: attendances[weekStart] })
+                    }
+                  >
+                    {formatDate(weekStart)}
+                  </div>
+                ))}
+              </div>
+              <table id="mentor-attendance-table">
+                <tbody>
+                  {selectedWeek &&
+                    stagedAttendances.map(({ id, student, presence }) => (
+                      <tr key={id}>
+                        <td>{student.name}</td>
+                        <td>
+                          <select
+                            value={presence}
+                            name={id}
+                            className="select-css"
+                            style={{
+                              backgroundColor: `var(--csm-attendance-${ATTENDANCE_LABELS[presence][1]})`
+                            }}
+                            onChange={this.handleAttendanceChange}
+                          >
+                            {Object.entries(ATTENDANCE_LABELS).map(([value, [label]]) => (
+                              <option key={value} value={value}>
+                                {label}
+                              </option>
+                            ))}
+                          </select>
+                        </td>
+                      </tr>
+                    ))}
+                </tbody>
+              </table>
+            </div>
+            <div id="mentor-attendance-controls">
+              <button className="csm-btn save-attendance-btn" onClick={this.handleSaveAttendance}>
+                Save
+              </button>
+              <button className="mark-all-present-btn" onClick={this.handleMarkAllPresent}>
+                Mark All As Present
+              </button>
+              {showSaveSpinner && <LoadingSpinner />}
+              {showAttendanceSaveSuccess && <CheckCircle color="green" height="2em" width="2em" />}
+            </div>
+          </React.Fragment>
+        )}
+        {!loaded && <h5> Loading attendances...</h5>}
+      </React.Fragment>
+    );
+  }
+}
+
+function MentorSectionInfo({ students, loaded, spacetime, override }) {
+  return (
+    <React.Fragment>
+      <h3 className="section-detail-page-title">My Section</h3>
+      <div className="section-info-cards-container">
+        <InfoCard title="Students">
+          {loaded && (
+            <table id="students-table">
+              <thead>
+                <tr>
+                  <th>Name</th>
+                </tr>
+              </thead>
+              <tbody>
+                {students.map(({ name, id }) => (
+                  <tr key={id}>
+                    <td>{name}</td>
+                  </tr>
+                ))}
+              </tbody>
+            </table>
+          )}
+          {!loaded && <h5>Loading students...</h5>}
```

Comment:
Would probably be better to use a ternary here in case of some weird race condition that causes both the `loaded` and `!loaded` branches to be rendered simultaneously; I know js is inherently single-threaded but I'd rewrite as a ternary just to be safe in case React or the DOM redraw things weirdly.

Reply:
So I was reading this really long blog post by Dan Abramov (one of the lead maintainers of React) on the `useEffect` hook (https://overreacted.io/a-complete-guide-to-useeffect/), and in the course of describing this he makes it very clear the React runtime guarantees any particular state/props values are tied to a _single_ concrete value during any one render. So we'll never see a mix of states. I try to stay away from the ternary operator for rendering when each option is more than a single line because I feel that it harms code readability, because you're hunting for this single tiny ':' to separate the two possible results.